File: src/Arqanore/Exceptions.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text;

namespace Arqanore
{
    public class ArqanoreException : Exception
    {
        public ArqanoreException(string message) : base(message)
        {
            
        }
        
        public ArqanoreException(string message, Exception inner) : base(message, inner)
        {
            
        }
    }

    public class ArqanoreShaderException : ArqanoreException
    {
        public ShaderType ShaderType { get; private set; }
        
        public ArqanoreShaderException(ShaderType shaderType, string message) : base(message)
        {
            ShaderType = shaderType;
        }

        public ArqanoreShaderException(ShaderType shaderType, string message, Exception inner) : base(message, inner)
        {
            ShaderType = shaderType;
        }
    }

    public class ArqanoreGlfwException : ArqanoreException
    {
        public int ErrorCode { get; private set; }
        public string ErrorDescription { get; private set; }
        
        public ArqanoreGlfwException(int errorCode, string errorDescription) : base("A GLFW error occurred.")
        {
            ErrorCode = errorCode;
            ErrorDescription = errorDescription;
        }

        public ArqanoreGlfwException(int errorCode, string errorDescription, Exception inner) : base("A GLFW error occurred", inner)
        {
            ErrorCode = errorCode;
            ErrorDescription = errorDescription;
        }
    }
}

File: src/arqanore/exceptions.cpp
#include "arqanore/exceptions.h"

int arqanore::GlfwException::get_code()
{
    return this->code;
}

arqanore::GlfwException::GlfwException()
{
    this->code = 0;
}

arqanore::GlfwException::GlfwException(int code, const char* message)
{
    this->code = code;
    this->message = message;
}

const char* arqanore::GlfwException::what()
{
    return message.c_str();
}

arqanore::ArqanoreException::ArqanoreException()
{
}

arqanore::ArqanoreException::ArqanoreException(const std::string& message)
{
    this->message = message;
}

const char* arqanore::ArqanoreException::what()
{
    return message.c_str();
}
